### PR TITLE
Fallback authentication for providers during DfE Signin outage

### PR DESCRIPTION
**Pull request by tijmenb**
## Context

We're experiencing a DfE Signin outage. This provides us with a way to turn on a fallback sign in mechanism for the provider interface.

## Changes proposed in this pull request

image

image

If you then click the link in your email, you've been signed in!

## Guidance to review

- Is this secure?
- Any edge cases?

## Link to Trello card

https://trello.com/c/t421tPrM/2073-deal-with-may-1st-dfe-signin-outage

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/views/provider_interface/sessions/sign_in_by_email.html.erb [hunk 1]
```diff
@@ -0,0 +1 @@
+<%= render 'candidate_interface/shared/check_your_email' %>
```
Comment (tvararu) [inline app/views/provider_interface/sessions/sign_in_by_email.html.erb:1]: This contains the `try again` link which links to the candidate sign in, right? Do we need to update that?

File: db/migrate/20191002123040_create_vendor_api_tokens.rb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-class CreateVendorApiTokens < ActiveRecord::Migration[6.0]
+class CreateVendorAPITokens < ActiveRecord::Migration[6.0]
```
Comment (tvararu) [inline db/migrate/20191002123040_create_vendor_api_tokens.rb:1]: Why is this needed?

File: app/controllers/provider_interface/sessions_controller.rb [hunk 3]
```diff
@@ -3,14 +3,58 @@ class SessionsController < ProviderInterfaceController
...
+
+      if provider_user
+        magic_link_token = MagicLinkToken.new
+        ProviderMailer.fallback_sign_in_email(provider_user, magic_link_token.raw).deliver_later
+        provider_user.update!(magic_link_token: magic_link_token.encrypted, magic_link_token_sent_at: Time.zone.now)
+
+        SlackNotificationWorker.perform_async(
+          "Provider user #{provider_user.first_name} has requested a fallback signin link",
+          edit_support_interface_provider_user_path(provider_user),
+        )
```
Comment (tvararu) [inline app/controllers/provider_interface/sessions_controller.rb:31]: Good idea

File: app/controllers/provider_interface/sessions_controller.rb [hunk 4]
```diff
@@ -3,14 +3,58 @@ class SessionsController < ProviderInterfaceController
...
+      end
+    end
+
+    def authenticate_with_token
+      magic_link_token = MagicLinkToken.from_raw(params.fetch(:token))
+      provider_user = ProviderUser.find_by!(magic_link_token: magic_link_token)
```
Comment (tvararu) [inline app/controllers/provider_interface/sessions_controller.rb:37]: Presumably we would check for token expiry at this stage. It doesn't look like we've built that yet which seems OK

File: db/migrate/20191002123040_create_vendor_api_tokens.rb (hunk 2 above)
Comment (tvararu) [reply db/migrate/20191002123040_create_vendor_api_tokens.rb:1]: As in, Duncan renamed these here: https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1937/files#diff-a0b8e3ce696609e3c61a95efee133caeR4

Comment (tijmenb) [reply db/migrate/20191002123040_create_vendor_api_tokens.rb:1]: The migrations can't run anymore because of https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1937. cc @duncanjbrown

File: app/views/provider_interface/sessions/sign_in_by_email.html.erb (hunk 1 above)
Comment (tijmenb) [reply app/views/provider_interface/sessions/sign_in_by_email.html.erb]: Fixed by copying over the content and changing the link.

File: app/controllers/provider_interface/sessions_controller.rb (hunk 4 above)
Comment (tijmenb) [reply app/controllers/provider_interface/sessions_controller.rb:37]: Added to https://trello.com/c/t421tPrM/2073-deal-with-may-1st-dfe-signin-outage.

File: config/routes.rb [hunk 5]
```diff
@@ -443,6 +443,8 @@
 
     get '/sign-in' => 'sessions#new'
     get '/sign-out' => 'sessions#destroy'
+    post '/sign-in-by-email' => 'sessions#sign_in_by_email', as: :sign_in_by_email
```
Comment (mnacos) [inline config/routes.rb:446]: Are these correct? The magic link email points users to `GET /provider/sign-in-by-email?token=325AB....`

Comment (mnacos) [reply config/routes.rb:446]: Review app gives me a 404
